gems/aws-sdk-cognitoidentity/lib/aws-sdk-cognitoidentity/customizations/cognito_identity_credentials.rb

```diff
@@ -0,0 +1,100 @@
+      #
+      # @option options [String] :identity_pool_id Required unless identity_id is
+      #   provided. An identity pool ID in the format REGION:GUID.
+      #
+      # @option options [Hash<String,String>] :logins A set of optional
+      #   name-value pairs that map provider names to provider tokens.
+      #   The name-value pair will follow the syntax
+      #   "provider_name": "provider_user_identifier".
+      #
+      # @option options [String] :custom_role_arn (optional) The Amazon Resource
+      #   Name (ARN) of the role to be assumed when multiple roles were received
+      #   in the token from the identity provider. For example, a SAML-based
+      #   identity provider. This parameter is optional for identity providers
+      #   that do not support role customization.
+      #
+      # @option options [Callable] before_refresh Proc called before
+      #   credentials are refreshed from Cognito.  Useful for updating logins/
+      #   auth tokens on expired
+      #
+      # @option options [Callable] token_expired proc called in
+      # near_expiration?.  If true, it will trigger a refresh and before_refresh
+      # will be called, allowing tokens in logins to be updated.
+      #
+      # @option options [STS::CognitoIdentity] :client Optional CognitoIdentity
+      #   client. If not provided, a client will be constructed.
+      def initialize(options = {})
+        @identity_pool_id = options.delete(:identity_pool_id)
+        @identity_id = options.delete(:identity_id)
+        @custom_role_arn = options.delete(:custom_role_arn)
+        @logins = options.delete(:logins) || {}
+        @before_refresh = options.delete(:before_refresh)
+        @tokens_expired = options.delete(:tokens_expired)
+
+        if !@identity_pool_id && !@identity_id
+          raise ArgumentError,
+                'Must provide either identity_pool_id or identity_id'
+        end
+
+        @client = options[:client] || CognitoIdentity::Client.new(
+          options.merge(credentials: false)
+        )
+        super
+      end
+
+      # @return [CognitoIdentity::Client]
+      attr_reader :client
+
+      # @return [Hash<String,String>]
+      attr_accessor :logins
+
+      # @return [String]
+      def identity_id
+        @identity_id ||= @client
+                         .get_id(identity_pool_id: @identity_pool_id)
+                         .identity_id
+      end
+
+      private
+
+      def refresh
+        @before_refresh.call(self) if @before_refresh
+
+        resp = @client.get_credentials_for_identity(
+          identity_id: identity_id,
+          custom_role_arn: @custom_role_arn
+        )
+
+        @credentials = Credentials.new(
+          resp.credentials.access_key_id,
+          resp.credentials.secret_key,
+          resp.credentials.session_token
+        )
+        @expiration = resp.credentials.expiration
+      end
+
+      def near_expiration?
+        (@tokens_expired && @tokens_expired.call(self)) ||
```

Comment:
The way I see it there's a one-way dependency here. Expired tokens affect the ability to get a new set of credentials, but don't affect the usability of credentials.

The concern I have with @tokens_expired is that it couples the token refresh with the creds refresh, forcing consumers to consider both when supplying this proc. I can see that causing confusion and unintended side-effects. That coupling would make sense if the token expiration was tied to the credential's expiration, but that's not the case.

It might be simpler and more flexible to just have @before_refresh as you have it above. It's executed every time the refresh is called, which is fine because the app can then choose how efficient/intelligent they want to be about refreshing their token within the supplied proc's logic.

Reply:
The case I was trying to address with this was a change in credentials - eg from the unauthenticated to authenticated state.  That would necessitate updating the token in the logins and refreshing the credentials.  Otherwise, you are correct that its a one way dependency and token refresh shouldn't be coupled with credentials refresh.  

I had originally wanted to handle the chase of changed credentials through something that would trigger a refresh of credentials when logins was updated.  Something like:
```ruby
       def logins=(logins)
        @logins = logins
        refresh!
      end
```

However, this doesn't handle the case of `credentials.logins['graph.facebook.com'] = new_token` and I couldn't easily find a way to support that aside from either 1) freezing the logins hash and requiring updates through `credentials.logins = {'graph.facebook.com'=>new_token}` or 2) having logins extend hash and include custom set logic w. a callback. 

Anyway - I do agree with you and will remove the `tokens_expired` and think about a solution for updates to logins.

Reply:
Oh, I see. My assumption was that updating a particular login's token would be done by the consumer inside the @before_refresh proc. Assuming of course that, like you have it, the CognitoIdentityCredentials object is available within that context so that people can modify the `logins` attr directly.

If you can find an elegant way to help facilitate that I'm game, but I think leaving it all to the user is totally acceptable.

Alternatively, if instead of, or in addition to, the generic @before_filter, there was a token-specific proc parameter with expected return value(s) - probably the entire logins hash - could you just set logins based on that proc's return value directly in the `refresh` function? As a bonus, if you could find a way to allow users to modify logins, without leaking other CognitoIdentityCredentials details that might be good. Maybe that's not worth the trouble though...

Reply:
After having thought through this a little bit more I've settled for keeping the fairly generic before_refresh proc as is that has the CognitoIdentityCredentials object available so that logins can be modified directly and in whatever way the user needs.  I've updated the documentation to reflect this and cover the case of how to update logins directly.

Reply:
Fantastic. I'm always a fan of simplicity.